### DISCUSSION TITLE
ScriptReference/SpriteMask

### DISCUSSION THREAD

**Discussion by docscomments** [Developer-Notes-Extension/Unity-Notes#318] [Notes]
# ScriptReference/SpriteMask

https://docs.unity3d.com/ScriptReference/SpriteMask.html

**Comment by cjacobwade** [comment]
NOTE: Using sprite mask and the maskInteraction setting on sprites affects the stencil value of these sprites when they're rendered even if that sprite's material has custom stencil settings.  
I found [this thread](https://forum.unity.com/threads/sprite-mask-inner-workings.492410/) explaining a bit better what's going on:  
- Short-ish version is when there's a sprite mask present this happens (going to simplify and pretend only sprites and masks are being rendered):  
- Sprite masks with custom ranges encapsulate all the sprites in those ranges  
- At the start of each group the sprite mask is rendered incrementing stencil value  
- Then all the sprites render and their MaskInteraction setting is used to set the stencil settings. Sprites set to VisibleInsideMask are set to draw if the existing stencil value is >= 1. Sprites set to VisibleOutsideMask are set to draw if existing stencil value is <1. Sprites set to none draw normally. This value is changed even if the spriteRenderer is using a material with a custom stencil setting  
- At the end of the transparent pass the sprite mask is rendered again decrementing the stencil value so later things aren't affected
